# День первый DWthon - выйди за рамки стандартного мышления!

### Цель на сегодня: загрузка и анализ исходных данных 

Добро пожаловать на **DWthon** - **выйди за рамки стандартного мышления** 🤝  Если ты чувствуешь, что немного потерялся в быстро меняющемся мире (все вокруг куда-то движутся и ты по инерции тоже) и хочешь из череды событий найти что-то перспективное и востребованное, но не знаешь с чего начать, то у меня для Тебя хорошая новость! Твой старт может стать эффективным и даже эффектным 🥳 **Тебе повезло**, т.к. *DWthon* - это правильное и нужное место 🧠. 

Привет! Меня зовут Владимир - я основатель DataWorkshop и автор строк, которые Ты сейчас читаешь. Моя роль - помочь Тебе раскрыть свой потенциал! Мир меняется. Причем очень быстро, но если "словить волну", то можно быть на ее гребне. Покажу Тебе, что это значит на практике и шаг за шагом вместе будем решать конкретную задачу, чтобы извлечь пользу из данных (за которую другие люди готовы платить). 

Анализировать данные можно вручную. Но когда их становится очень много (особенно если речь идет про "большие данные" - англ. *big data*)- это, как правило, мало эффективно, поэтому нужно использовать другие инструменты, например, искусственный интеллект. Хотя, говоря точнее, будем обучать машину так, чтобы она нам помогала - это занятие так и называется - машинное обучение 👀.     

### Чего можешь ожидать?
*DWthon* длится (всего лишь) 5 дней и поэтому нет времени распыляться на мелочи, вдаваясь в мало практические подробности (различные хитрые формулы), в машинном обучении. Задача в том, чтобы показать Тебе, как сделать первые шаги в машинном обучении, получить конкретный результат (рабочее решение) и научить иному образу мышления. При случае, конечно же, еще вдохновить (без этого трудно двигаться дальше) и начать раскрывать Твой потенциал (а раз Твоя судьба занесла Тебя сюда, то, наверное, это не случайно!). **Когда чего-нибудь сильно захочешь, вся Вселенная будет способствовать тому, чтобы желание твое сбылось** (с) - пожалуйста, не упусти этот шанс! ☝️


### Черные дыры
Возможно, Тебе приходилось слышать про черные дыры в космосе, но я сейчас имею в виду нечто другое. Объясню. В DataWorkshop мы говорим так: "Будь осторожен и обходи стороной черные дыры!" Что это значит? В мире, где мы живем, существует множество задач, которые "съедают" Твою жизненную энергию, внимание, время, деньги и другие ресурсы, а взамен не дают ничего. Это как пылесос, который готов  бесконечно поглощать Твои ресурсы. Как часто у Тебя были такие дни, когда целый день что-то делаешь и делаешь, под конец дня чувствуешь себя очень уставшим, но в тоже время понимаешь - ничего важного сегодня не сделано. И на вопрос: "Так чем я сегодня в итоге занимался?" порой ответить не так легко (а иногда и не хочется). Помнишь такие дни и задания когда "черная дыра" ловила Тебя?

###  Наука о данных
Так вот, хочу Тебе сразу сообщить, что в машинном обучении и в науке о данных (англ. `Data Science`) - просто полно черных дыр! Здесь их так много, что нужно быть очень внимательным. На практике это значит, что очень легко быть занятым и выполнять много никому ненужной работы. Эффект нулевой (и даже отрицательный), т.к. время и другие ресурсы потрачены. И это неправильно, так не должно быть!  Именно поэтому, я, как практик, хочу заострить Твое внимание на этом.

С одной стороны я хочу научить Тебя создавать рабочие решения, используя науку о данных (в том числе машинное обучение), но в то же время предупредить, что этот мир с данными очень коварный и с радостью "засасывает" новичков в черные дыры, выжимая из них все соки! Предупреждён — значит вооружён 💪. Но предупредить мало. Вот почему был создан *DWthon*, я Тебе покажу шаг за шагом - что имею в виду и как с этим всем жить и справляться.


###  Закон Парето
DataWorkshop - это прежде всего практика. Поэтому, если слышишь от нас закон с каким-то умным названием - значит в нем есть практика. И на DWthon увидишь - как это можно применять!

Объясняя материал (машинное обучение или другое), я всегда стараюсь использовать человеческий язык, чтобы не прятаться за сложными формулировками. Но стоит заметить, что зависимость - меньшинство дает большинство и наоборот, заметили уже давно и даже этому дали свое название. Существует **закон Парето - 80/20**, который говорит о том, что 80% чего-либо (здесь Ты можешь вставить почти все, что захочешь, например, стоимость, энергия, потери и т. д.) генерируется 20% чего-либо (например, усилия, клиенты и т.д.). 

Изначально речь шла про финансовый аспект. Человек заметил, что 80% богатств принадлежит 20% людям (сейчас эта пропорция еще более нарушена).


## Низко висящие фрукты
😱 Наша задача - искать низко висящие фрукты! Зачем тянуться и неэффективно срывать "фрукты", которые находятся дальше. Гораздо лучше задействовать потенциал, который доступен здесь и сейчас.  Этим и займемся!

Будем учиться искать и высматривать👀 наиболее важные задачи, чтобы под конец DWThon получить классный результат. Такой результат, который дает конкретную ценность (выгоду) и кто-то готов за это платить.

Наконец, еще одна очень важная вещь, и будем начинать. У меня есть просьба 🙏 Не сравнивай себя с другими. На *DWthon* все - победители. Почему? Сам факт того, что Ты здесь, стараешься и пробуешь - это уже круто! Первый шаг сделан (читаешь), теперь пора сделать следующий шаг. Обращай внимание на каждый твой *DWthon* день. Что почерпнул нового - сравнивай с предыдущими днями. Измеряй, пожалуйста, только свою "дельту" развития, а не других ☝️. Так будет эффективнее! И у Тебя будет чувство удовлетворенности от быстрого продвижения вперед 🚀.


### Внимание! 
Ты учишься - и у Тебя могут быть вопросы. Это нормально! Мало того, это даже очень хорошо. Спрашивая, Ты более глубоко начинаешь лично прощупывать тему. Поэтому, помни если у Тебя возникли какие-либо трудности или сомнения, при выполнении задач, то обязательно напиши об этом в Slack (это наш "чатик" для общения). Все вопросы, которые касаются первого дня, стоит писать в комнате [dwthon_day1](https://bit.ly/2UTWe1G)

Не забудь подробно указать, в чем именно сложность. Лучше всего выложить "скриншот" (фотку своего экрана) со своим кодом, где видна проблема. Тем самым поможешь другим людям понять, что у Тебя происходит и они тогда смогут Тебе помочь!

Также старайся помогать другим. Благодаря этому, мы все вместе будем продвигаться вперед.


## Ну что... Поехали! 🛼
Сначала посмотри видео ниже.

Подсказка. Чтобы его увидеть, перемести курсор в ячейку с видео и нажми одновременно `SHIFT` +` ENTER`.

In [1]:
%%html
 <iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/UkZPOYMczUQ" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## Видео - шаг за шагом

Чтобы еще больше Тебя поддержать, я дополнительно записал "пошаговое" видео. Советую его использовать, если Твое путешествие в IT только начинается - там будет больше комментариев по выполнению заданий. Это видео поможет "переварить" полученную информацию.

Кстати, помни о том, что мозгу человека необходимо время, чтобы (совсем) новая информация усвоилась и образовались нужные нейронные соединения. Поэтому, если что-то непонятно - это нормально, просто продолжай идти вперед. Главное идти. Точки объединятся, только позволь себе это (дай немного времени на это)!

In [2]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/PeFYE3XwqOY" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## Python

Начинаем кодить, т.е. программировать 👌🤩🥳.

Мы будем использовать язык программирования Python. Произносится это как "пайтон" (так звучит по английски), хотя по "старинке" в русскоязычных странах его часто называют "питон". Это достаточно простой язык, но очень полезный. На питоне можно делать очень много, и между прочим, создавать решения для машинного обучения. Кстати, по некоторым оценкам питон является языком  #1 по популярности. Хотя это не точно, все зависит от того - кто и как это считает, но однозначно можно сказать, что этот язык входит в TOP5 языков, которые полезны и дают возможность зарабатывать!


Для упрощения и ускорения разработки существует множество ранее реализованных разработок. Для того, чтобы был порядок, существует определенная иерархия: модуль, пакет, библиотеки....

- **модуль** (англ. `module`) - это файл, содержащий различные функции Python и глобальные переменные. Это просто файл с расширением .py с исполняемым кодом Python.
- **пакет** (англ. `package`) - это набор модулей. Он должен содержать файл init.py в качестве флага, чтобы интерпретатор python обрабатывал его как таковой. 
- **библиотека** (англ. `library`) - это набор пакетов.


## Библиотека

Сегодня мы познакомимся с первой библиотекой - `pandas`. Эта библиотека поможет работать с данными, для начала импортировать данные и посмотреть - что там внутри (какие столбцы, значения и т.д.).

Для упрощения можно сказать, что `pandas` предоставляет множество возможностей, аналогичных excel 🥳, только здесь мы пишем код с помощью **языка Python**. 

Для того, чтобы начать работать с библиотекой, ее нужно загрузить (проимпортировать ее логику). Это делается очень просто, используя ключевое слово `import`.

## Как запустить (выполнить) ячейку?
* `Shift + Enter` - выполняет код данной ячейки и переходит к следующей ячейке. 

☝️ Помести курсор в ячейку с приведенным ниже кодом и нажми `Shift + Enter`. Если перед ячейкой слева появится число в [...], то это успех - код "выполнен". Выполни ту же процедуру в каждой последующей ячейке, где видишь код. 

In [3]:
import pandas as pd

`pd` - это псевдоним ` pandas` (делается, как правило, потому, чтобы меньше букв писать - вместо слова `pandas` будем писать всего лишь две буквы `pd`).


## ✔️ Загружаем данные

Мы берем данные (в виде таблицы) из файла `data.csv` (`csv` - это такой упрощенный формат `excel`).

Для загрузки данных мы используем функцию из `pandas` под названием `.read_csv()`. В качестве аргумента мы указываем путь к файлу (путь, по которому файл с данными физически хранится на диске). 

В нашем случае файл хранится в папке `input` и имеет имя `data.csv`. Мы загружаем и присваиваем результат переменной `df`. На практике это означает, что, вызывая `df`, мы получаем доступ к загруженному набору данных.

Одной строки кода достаточно, чтобы загрузить почти миллион строк в секунду. 👍

In [4]:
df = pd.read_hdf("../input/data.h5")

## Таблица

☝️ Интересный факт: *В терминологии `pandas` мы называем таблицу - `dataframe` (отсюда и название переменной `df`).

Давай проверим "размер" загруженной нами таблицы, используя для этого `.shape`. Когда мы говорим о размере таблицы, мы имеем в виду количество строк и столбцов.

In [5]:
df.shape

(820906, 9)

Таблица содержит:
- **820 906** строк 
- **9** столбцов

Мы уже знаем основную информацию о нашей таблице. Теперь давай посмотрим на данные, например, на первые 5 строк, используя функцию `.head()`.
Ну что, идем дальше? 😉

In [6]:
df.head()

order_id  customer_id  product_id  quantity  price_unit  price_total  \
0         0            0           0         6         254         1524   
1         0            0           1         6         339         2034   
2         0            0           2         8         275         2200   
3         0            0           3         6         339         2034   
4         0            0           3         6         339         2034   

   country_id          order_date  is_canceled  
0           0 2010-12-01 08:26:00        False  
1           0 2010-12-01 08:26:00        False  
2           0 2010-12-01 08:26:00        False  
3           0 2010-12-01 08:26:00        False  
4           0 2010-12-01 08:26:00        False

## Краткая информация

- `pd.head()` - возвращает нам первые строки. По умолчанию он возвращает первые 5 строк (из нашей таблицы).
- `pd.sample()` - вернет случайные строки, по умолчанию - одна (`pd.sample (10)` вернет 10 случайных строк)

Обе функции полезны, хотя стоит признать, что просмотр случайных строк (запуск ячейки, например, 5-10 раз) может принести гораздо больше пользы. Дает шанс увидеть "что-то" интересное внутри, а не только смотреть на первые пять результатов. Иногда случайные строки называют "рандомные" или "рандомными" от английского слова `random` (что означает: случайный или произвольный).

Попробуй запустить ячейку ниже (там, где код `df.sample(10)`) 5-10 раз. Посмотри внимательно на значения, которые появляются. Попробуй "уловить" определенные зависимости. Потрать, пожалуйста, 1-3 минуты на это задание.


### 💡 Подсказка
Обычно мы запускаем ячейку `SHIFT` + `ENTER`. В результате ячейка запускается и "перескакивает" на следующую. Иногда нам просто хочется запустить и оставаться в той же самой ячейке (например, сейчас так хочется). Тогда разумнее использовать `CTRL` +` ENTER` - ячейка выполняется, но курсор остается в той же самой ячейке. Попробуй 😉.

In [16]:
df.sample(10)

order_id  customer_id  product_id  quantity  price_unit  price_total  \
489226     31559         4889          61         6         375         2250   
110354      7776         2379         100        12         145         1740   
82122       5722         1972        1170         1         495          495   
61340       4401         1046        1248         4         375         1500   
295830     19598         3821         608         1         329          329   
57587       4130         1546         255        10         165         1650   
721221     47239          265          48         6         125          750   
366658     23225         2737         240        10         125         1250   
452616     28678         1154          62        12          85         1020   
292130     19430         1091        1461        12         210         2520   

        country_id          order_date  is_canceled  
489226           0 2010-03-07 10:55:00        False  
110354           0 2011-04-17 10:21:00        False  
82122            0 2011-03-17 12:47:00        False  
61340            0 2011-02-20 12:23:00        False  
295830           0 2011-10-16 14:35:00        False  
57587            0 2011-02-15 10:40:00        False  
721221           0 2010-10-17 14:30:00        False  
366658           0 2011-11-21 16:25:00        False  
452616           0 2010-01-25 12:08:00        False  
292130          18 2011-10-13 12:47:00        False

Это цифры для интернет-магазина из Великобритании. Давай разберемся с контекстом. 

Я расскажу Тебе историю, чтобы стимулировать воображение. Это поможет лучше и быстрее понять данные. А это всегда очень важно. Помни, что всегда полезно знать контекст и источник данных, с которыми имеем дело. 

История обычно очень помогает, тогда нам легче смотреть на данные, чем просто на "сухую" таблицу. Наш мозг любит слушать истории, так давай поможем ему :) 


## 🏪 Пойми проблему - как работает магазин? 

Давай начнем с вопроса: **как работает обычный магазин?

Есть клиенты (`customer_id`), которые покупают товары (`product_id`) по цене (`price_unit`) в количестве (`quantity`), и все это объединяется в заказ (`order_id`). 

Кроме того, у нас есть время, когда была совершена покупка `order_date`. 

Следует помнить, что покупатель имеет право вернуть заказ. Если он воспользуется этим правом, то поле `is_canceled` будет установлено в `True`. В любом другом случае `is_canceled=False`, т.е. возврата не произошло.



### ☝️ Примечание - мы используем ID.

*С точки зрения удобства (и не только), выгоднее хранить уникальный идентификатор (т.е. `id` - заимствовано из английского `identifier`), чем, например, непосредственно **имя** и **фамилия**. 

Это удобно, потому что имя и фамилия слишком подвержены опечаткам или лишним символам (например, слишком много пробелов), или орфографическим неточностям, например, одна буква в верхнем регистре, другая в нижнем. Стоит помнить, что данные также меняются, например, может измениться фамилия. Стоит помнить и политике конфиденцальности. Имя и фамилия являются частными данными. Внутренний `customer_id` не обязательно является таковым, но не воспринимай это как юридический совет. Это лишь совет, всегда стоит обсуждать такие вопросы с юристами и чиновниками 😎)!*. 


#### Пример

Числа и названия случайны. Я использую этот пример, чтобы стимулировать Твое воображение.

Маша (`customer_id=1`) делает заказ (`order_id=123`) 8 марта в 10:00:00 (`order_date=03-08-2021 10:00:00`), который состоит из следующих продуктов:
- **хлеб** (`product_id=1`) одна штука (` amount=1`) за 3 рубля (`price_unit=300`)
- **помидоры** (`product_id=2`) 5 штук (`amount=5`) всего по 2,55 рублей / шт (`price_unit=255`) **2,55* 5 = 12,75 рублей** (` price_total=1275`)
  
  
Кстати, стоит заметить, что Ты, как потребитель, имеешь право на возврат продуктов в течение некоторого времени (в Америке и Европе этим часто пользуются). В данном случае данные так записаны, что если возвращаешь даже один продукт - это означает возврат всего заказа, а не только одного продукта (это просто так кто-то выдумал, даже не трать особо времени - почему, кто-то выдумал что хочет именно так 😂). В наших данных заказы, которые были возвращены, будем обозначать как: `is_canceled=True`. В этом случае мы говорим, что заказ отменен (был возврат). Кстати, при возврате магазин полностью отдает деньги за заказ.


### ☝️ Обрати внимание, как мы сохраняем «цену» в рублях.

#### 12,75 рублей - записываем как "1275" - почему?

*делаем это по простой причине, чтобы не нужно было беспокоиться об округлении чисел. Мы пытаемся вдохновиться, как это работает на «физическом уровне» нашего мира. Самая маленькая денежная единица - «копейки» (цент и т.д.). Поэтому давай использовать это мышление и здесь, и мы сэкономим себе много времени и нервов (если хочешь, можно почитать  подробнее о проблеме - как "записывать деньги" в цифровом формате [здесь](https://bit.ly/2PI9oMt). Но помни наш подход - 80/20, т.е. мы пытаемся создавать 80% ценности при 20% усилий, поэтому давай не мучаться и пользоваться очень простым механизмом, тем самым упростим себе жизнь.*

## С чего начать анализ данных?

Как думаешь, если работаешь с бизнес-проектом, то с чего нужно начать? Ответ почти всегда один и тот же, вне зависимости от культуры и времени - «деньги» 💰. Это очень простое правило, с помощью которого Ты можешь давать ценность (помогать зарабатывать или хотя бы экономить деньги) компаниям и, благодаря этому, зарабатывать до **10 000 $** или больше в месяц. Немного об этой теме уже говорил в [видео](https://youtu.be/yUcbr0ZlEu4).

Способность приносить пользу «бизнесу» тоже интересна по другой причине. Благодаря этому работодатель будет искать именно Тебя, а не наоборот (это весьма удобно). Улови разницу! Наверное, можешь подумать, что я шучу, но, тем не менее, я говорю правду 😉. Можно даже сказать,что призываю организовать свою жизнь так, чтобы именно Ты был нужен многим компаниям (а не наоборот). Я постараюсь Тебе в этом помочь или хотя бы расскажу, как это у меня работает 🤝. Приступим к анализу данных ;)

**Выручка**

Посчитаем, сколько денег заработал магазин. Почему это важно? Много причин, хотя бы потому, что если магазин генерирует, например, 100 рублей, то даже если мы придумаем очень блестящую систему улучшения, что с того? 

С другой стороны, если магазин генерирует 10 миллионов и более, то изменения в один процент - это уже много (100 000, не так ли?). Если делать 10 таких изменений в год, то твоя зарплата ~ $10к в месяц уже не звучит как что-то волшебное. Ты можешь увидеть причину и следствие (такой зарплаты).

Повторюсь, но давай придерживаться правила 80/20. Делать наиболее важные вещи - все остальное оставляем на потом. Нужно сосредоточиться на том, чтобы делать то, что действительно оказывает значительное влияние. Направляя свою жизненную энергию в нужное русло, мы достигаем гораздо больше.

Так, смотрим на обороты магазина.

In [17]:
df["price_total"].sum()

1816145630

`price_total` это комбинация `price_unit` * `quantity` для данного продукта.
Другими словами, нахождение суммы (функция `.sum()`) для `price_total` помогает получить информацию о выручке этого магазина. ☝️ 

Напомню, что `price_total` (а также `price_unit`) держим в «копейках». Собственно, давай сразу же добавим контекст. Магазин, который мы анализируем, находится в Великобритании и их деньги - фунты стерлингов. Один фунт это 100 пенсов (так называются их "копейки"). Другими словами `price` держим в "копейках", которые в данном случае называем "пенсы".

#### Хорошо, сколько там вышло?

Если присмотришься 👀, то увидишь, что ... **18 161 456,30**, что составляет 18+ миллионов фунтов стерлингов.🧐. Похоже, у нас есть что оптимизировать - это очень хорошо! 🥳

Теперь проверим, сколько денег ускользнуло (т.е. возвраты клиентов). В этом случае мы оставляем только строки "возвраты", применяя простой фильтр `is_canceled = True`, а затем вычисляем сумму. Смотри, как это легко и интуитивно делается в `pandas`.

In [18]:
df[ True == df.is_canceled  ]["price_total"].sum()

71973973

Оказалось, что магазину нужно было вернуть клиентам достаточно "круглую сумму" **719 739,73 фунтов стерлингов**. Но мы то радуемся и говорим - ура! Откуда такая радость? 🤦‍♂️ Потому что есть что оптимизировать или сэкономить.

Давай теперь изучим еще несколько вещей. Важные параметры:
- сколько у нас клиентов?
- сколько у нас продуктов?

В `pandas` у нас есть функция `nunique()` (☝️ обрати внимание в начале на `n`), эта функция возвращает количество уникальных значений. Посмотрим на это в действии.

In [19]:
df["customer_id"].nunique()

5879

Видим **5879**, это означает, что в столбце `customer_id` столько уникальных значений. У каждого клиента есть свой идентификатор - `customer_id`. Таким образом, уникальное количество значений в столбце `customer_id` отвечает на вопрос: "сколько в магазине клиентов?"

Понятно, правда?

## Задание 1.1

Для Тебя уже понятно, как определить количество клиентов. Теперь нужно поработать немного самостоятельно и сделать то же самое для продуктов. Мы хотим ответить на вопрос, сколько уникальных продуктов (товаров) имеется в ассортименте магазина?


<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть подсказку 👈 </summary>
<p>
    <h3>💡 Подсказка</h3>

1. Вместо содержания `## YOUR CODE HERE` 👇👇 напиши свой код.
2. Вдохновись ячейкой выше...     
3. Скорее всего, Ты захочешь использовать столбец `product_id` 😉, а выше мы нашли уникальное количество покупателей, используя столбец `customer_id`.

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть ответ 👈 </summary>
<p>
    
```python
df["product_id"].nunique()
```
Старайся заглядывать в эту секцию только после того, как самостоятельно сделаешь задачу, чтобы сверить свое решение с нашим. 😉
</p>
</details> 
</p>
</details>

In [25]:
## YOUR CODE HERE
df["product_id"].nunique()

3878

### 💡 Дополнительная подсказка
Чтобы вставить пустую ячейку для ввода своего кода, можно воспользоваться горячими клавишами:
1. **CTRL** + **M** + **A**  - вставить ячейку сверху (Above)
2. **CTRL** + **M** + **B**  - ячейка снизу (Below)

## Задание 1.2
Поскольку мы уже знаем, сколько у нас товаров, давай проверим, **сколько у нас заказов** в магазине.

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть подсказку 👈 </summary>
<p>
    <h3>💡 Подсказка</h3>

1. Вместо `## YOUR CODE HERE` напиши свой код
2. Обрати внимание на столбец `order_id`.

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть ответ 👈 </summary>
<p>
    
```python
df["order_id"].nunique()
```
Старайся заглядывать в эту секцию только после того, как самостоятельно сделаешь задачу, чтобы сверить свое решение с нашим. 😉
</p>
</details> 
</p>
</details>

In [26]:
## YOUR CODE HERE
df['order_id'].nunique()

43908

Это был план минимум. Чтобы задание было засчитано за выполненное, нужно:
1. Сделать скриншот, на котором видны решения заданий 1.1 и 1.2
2. Поместить скриншот на канал [dwthon_day1_done](https://bit.ly/3iaa6xL)
3. Получить бонус за хорошо выполненную работу 💪 (кстати, найди время чтобы поработать с материалом, тем самым быстрее пойдешь вперёд) и конечно же завтра будет следующая порция знаний.

### Вопросы?
Напомню. Если у Тебя возникла проблема с задачей или чего-то не понимаешь, помни, что Ты всегда можешь спросить! Чего-то не знать - это нормально и спрашивать - это хорошо :). Задавай вопросы, относящиеся ко дню DWthon1, в канале [dwthon_day1](https://bit.ly/2UTWe1G). Также, если можешь помочь другим ответить на возникшие вопросы в чате, то сделай это, пожалуйста :).

## Дополнительные задания

Эти задания необязательно делать, но когда есть желание поделать еще что-то с данными, то обязательно продолжай 👍.


### 🧠 Включим критическое мышление.
Этот навык очень необходим при анализе данных (и вообще пригодится в жизни).

Выше была задача - сколько товаров в магазине, верно? Но можем ли мы ответить на этот вопрос? Давай подумаем...

На самом деле нет! Потому что мы не знаем, сколько товаров в магазине. Что мы знаем - это какие товары продавались хотя бы раз! Это огромная разница, потому что может быть тысяча других продуктов (которых нет в наших данных / заказах), которые генерируют затраты, но не продаются вообще. Поэтому внимательно старайся "выхватывать" контекст. Важно оперировать фактами (не домыслами).

Теперь посчитаем «средний чек». Мы знаем, какая выручка. Мы ведь тоже знаем - сколько заказов? Знаем! Тогда давай разделим одно на другое и получим информацию о "среднем чеке".

In [27]:
df["price_total"].sum() / df["order_id"].nunique()

41362.5223193951

Мы так легко узнали, что средний чек составляет 413,63 фунта стерлингов. Довольно много, правда?

Подскажу, что в данном случае это сувенирный магазин. Возможно, вспомнишь тоже, что когда делаешь новогодние покупки-подарки для своей семьи или друзей, то, вероятно, тоже тратишься - больше чем обычно.


Хорошо, давай немного подумай самостоятельно.

## 😇 Включим творческое мышление

Какие еще 5-10 показателей (кроме "среднего чека") стоит проверить? Время на креатив, спокойно подумай и Ты обязательно что-нибудь выдумаешь интересное. Я в это верю 👍.

Тебе просто нужно найти пространство, где можно сосредоточиться и подумать, может быть, небольшая прогулка, быстрый холодный душ 🚿, медитация или что-то еще, что побудит мыслить творчески.

Например. Спроси себя так. Я ответственный за интернет магазин, какие показатели меня должны волновать больше всего и почему? Но можешь пойти дальше. 



Подброшу пару важных метрик, чтобы поставить Твое мышление на правильные рельсы.

1. **Средний чек** (англ. `Average Order Value` или `AOV`): `AOV` - это средняя сумма денег, которую клиент тратит за один заказ. Важно отслеживать эту метрику, потому что она помогает понять, сколько выручки вы можете получить от каждого клиента. 
2. **Жизненная ценность клиента** (англ. `Customer Lifetime Value ` или `CLV`): `CLV` - это общая сумма денег, которую клиент потратит на товары  в магазине за свою жизнь (для упрощения обычно берется какой-то обозримый период времени, например в течение 1-2 лет). Эта метрика важна, потому что она помогает понять долгосрочную ценность клиента и сколько вы можете позволить себе тратить на его привлечение.
3. **Коэффициент возврата** (англ. `Return Rate`): Это процент клиентов, которые возвращают продукт, который они купили. Важно отслеживать эту метрику, потому что это может помочь выявить проблемы с качеством продукта или обслуживания клиентов.
Частота повторных покупок: Эта метрика сообщает вам, как часто клиенты возвращаются за покупками на ваш сайт. Это может помочь вам понять, насколько удовлетворены клиенты вашими продуктами и обслуживанием, а также помочь определить потенциал для увеличения продаж.
4. **Количество продаж**: Это количество единиц продукта, которые были проданы за определенный период времени. Эта метрика позволяет оценить популярность продукта и его спрос среди покупателей. 
5. **Частота повторных покупок**: Эта метрика сообщает, как часто клиенты возвращаются за покупками в магазин. Это может помочь  понять, насколько удовлетворены клиенты продуктами (и обслуживанием), а также помочь определить потенциал для увеличения продаж.


Пробуй! Экспериментируй, в этом смысл данного задания - находить менее очевидные на первый взгляд вещи (хотя когда уже нащупаешь их, то они кажутся очень логичными).

In [33]:
df

order_id  customer_id  product_id  quantity  price_unit  price_total  \
0              0            0           0         6         254         1524   
1              0            0           1         6         339         2034   
2              0            0           2         8         275         2200   
3              0            0           3         6         339         2034   
4              0            0           3         6         339         2034   
...          ...          ...         ...       ...         ...          ...   
820901      1070          620         362         2         295          590   
820902      1070          620         395         1         375          375   
820903      1070          620         364         1         375          375   
820904      1070          620        1242         2         375          750   
820905      1070          620          73         2         195          390   

        country_id          order_date  is_canceled  
0                0 2010-12-01 08:26:00        False  
1                0 2010-12-01 08:26:00        False  
2                0 2010-12-01 08:26:00        False  
3                0 2010-12-01 08:26:00        False  
4                0 2010-12-01 08:26:00        False  
...            ...                 ...          ...  
820901           0 2010-12-09 20:01:00        False  
820902           0 2010-12-09 20:01:00        False  
820903           0 2010-12-09 20:01:00        False  
820904           0 2010-12-09 20:01:00        False  
820905           0 2010-12-09 20:01:00        False  

[820906 rows x 9 columns]

### Средний чек (англ. Average Order Value или AOV)

In [56]:
df["price_total"].sum() / df["order_id"].nunique()

41362.5223193951

### Жизненная ценность клиента (англ. Customer Lifetime Value)

In [58]:
df.groupby(['customer_id'])["price_total"].sum() / df["order_id"].nunique()

customer_id
0       130.879817
1        20.727726
2        30.460599
3         5.155803
4         7.902387
           ...    
5874      1.101007
5875      0.685092
5876      2.525007
5877      0.306505
5878      0.049877
Name: price_total, Length: 5879, dtype: float64

### Коэффициент возврата (англ. Return Rate)

In [54]:
# return rate
df['order_id'].nunique() / df[True == df.is_canceled]["is_canceled"].sum()
# частота покупок

2.4470824276876777

### Количество продаж

In [35]:
df.groupby(['customer_id'])['quantity'].sum()

customer_id
0       22976
1        3514
2        8971
3        1108
4         390
        ...  
5874      213
5875      261
5876      367
5877       85
5878        2
Name: quantity, Length: 5879, dtype: int32

## 🤝🗣️ Совместное обсуждение и обмен опытом

Поделись, пожалуйста, своими мыслями с другими. Зачем? Благодаря этому могут родиться еще более качественные (совместные) идеи. Это когда 1+1 = 11 😉

Напиши в чатике [dwthon_day1_ideas](https://bit.ly/3zIbBsN), что у Тебя получилось выдумать. Поучаствуй в обсуждении, и увидишь на примере - как легко генерировать множество идей в сотрудничестве с другими. Это очень полезный навык, который скорее всего Тебе пригодится в современном мире.

## Отзыв и дельта Твоего прогресса

Дай мне, пожалуйста, знать, что Тебе понравилось, было ли трудно (и где именно) и зафиксируй также для себя - чему получилось научиться на этом уроке.

### Тут 👉👉 [анкета](https://bit.ly/3j0lLyq) 👈👈
Твой отзыв позволит улучшить материал!

## До завтра! 🙌